src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
+        {
+            _schema = schema;
+            _typeToPropertyValueConstructorMap = primitiveValueBuilders ?? GetBuildersForRequiredPrimitives();
+        }
+
+        private JsonSchema _schema;
+        private IDictionary<Type, PrimitiveValueBuilder> _typeToPropertyValueConstructorMap;
+
+        // There are two places in the format where object nest references to the defining 
+        // type. 'exception.innerExceptions' is an array of exception objects. 
+        // 'node.children' is another collection of nodes. We need to watch for these
+        // scenarios to prevent re-entrancy and eventual stack consumption when 
+        // producing samples of these types.
+        private int graphNodeParentCount;
+        private int exceptionDataParentCount;
+
+        public override object VisitActual(ISarifNode node)
+        {
+            PopulateInstanceWithDefaultMemberValues(node);
+
+            return base.VisitActual(node);
+        }
+
+        // Retain nesting level for visiting exceptions to prevent
+        // unbounded re-entrance populating exception.innerExceptions
+        public override ExceptionData VisitExceptionData(ExceptionData node)
+        {
+            exceptionDataParentCount++;
+            node = base.VisitExceptionData(node);
+            exceptionDataParentCount--;
+
+            return node;
+        }
+
+        // Retain nesting level for visiting graph nodes, to prevent
+        // unbounded re-entrance populating exception.innerExceptions
+        public override Node VisitNode(Node node)
+        {
+            graphNodeParentCount++;
+            node = base.VisitNode(node);
+            graphNodeParentCount--;
+
+            return node;
+        }
+
+        private void PopulateInstanceWithDefaultMemberValues(ISarifNode node)        {
+           
+            Type nodeType = node.GetType();
+            var binding = BindingFlags.Public | BindingFlags.Instance;
+            foreach (PropertyInfo property in nodeType.GetProperties(binding))
+            {
+                // The node kind is always properly set in the OM and
+                // isn't relevant to the SARIF schema itself
+                if (property.Name == "SarifNodeKind") { continue; }
```

Comment:
>SarifNodeKind [](start = 38, length = 13)

Huh. So the SDK introduces a _de facto_ reserved property that the spec doesn't mention. Do we want to do what we've discussed in the past, that is, reserving property names that start with "Sarif" (any case) to the implementation?

Reply:
Hm. Let's discuss offline.

---
In reply to: [229111818](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229111818) [](ancestors = 229111818)

Reply:
I'm wrong. This is not a member of the property bag (_e.g._, `physicalLocation.Properties["SarifNodeKind"]`). Rather, it is a first-class .NET property of the generated class (_e.g._, `physicalLocation.SarifNodeKind`). Forget I said anything.

---
In reply to: [229116171](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229116171) [](ancestors = 229116171,229111818)